File: admin/src/api/demo/usage.ts
import type { DemoRoute } from './types'
import { getStore, generateId, nowISO, daysAgo } from './store'
import type { UsageLimitData } from './store'

const defaultLimits: UsageLimitData[] = [
  { id: 1, service_type: 'llm', limit_type: 'daily', limit_value: 1000, warning_threshold: 80, hard_limit: false, enabled: true, created: daysAgo(30) },
  { id: 2, service_type: 'tts', limit_type: 'daily', limit_value: 500, warning_threshold: 80, hard_limit: false, enabled: true, created: daysAgo(30) },
  { id: 3, service_type: 'llm', limit_type: 'monthly', limit_value: 20000, warning_threshold: 80, hard_limit: true, enabled: true, created: daysAgo(30) },
]

function generateDailyData() {
  const daily = []
  for (let i = 29; i >= 0; i--) {
    const d = new Date()
    d.setDate(d.getDate() - i)
    daily.push({
      date: d.toISOString().split('T')[0],
      units: Math.floor(50 + Math.random() * 200),
      requests: Math.floor(20 + Math.random() * 80),
      cost_usd: +(Math.random() * 2).toFixed(3),
    })
  }
  return daily
}

export function initUsageData() {
  const store = getStore()
  if (store.usageLimits.length === 0) {
    store.usageLimits = [...defaultLimits]
  }
}

export const usageRoutes: DemoRoute[] = [
  {
    method: 'GET',
    pattern: /^\/admin\/usage\/logs/,
    handler: () => {
      const logs = []
      for (let i = 0; i < 20; i++) {
        logs.push({
          id: i + 1,
          timestamp: daysAgo(Math.floor(Math.random() * 7)),
          service_type: ['llm', 'tts', 'stt'][Math.floor(Math.random() * 3)],
          action: ['chat', 'synthesis', 'transcribe'][Math.floor(Math.random() * 3)],
          source: ['admin', 'telegram', 'widget'][Math.floor(Math.random() * 3)],
          units_consumed: Math.floor(10 + Math.random() * 500),
          cost_usd: +(Math.random() * 0.5).toFixed(4),
        })
      }
      return { logs, count: 20 }
    },
  },
  {
    method: 'GET',
    pattern: /^\/admin\/usage\/stats/,
    handler: () => ({
      stats: {
        period_days: 30,
        from_date: daysAgo(30),
        by_service: {
          llm: { units: 15420, requests: 892, cost_usd: 12.34 },
          tts: { units: 8730, requests: 456, cost_usd: 4.56 },
          stt: { units: 2340, requests: 123, cost_usd: 1.23 },
        },
        by_source: {
          admin: { units: 3200, requests: 180 },
          telegram: { units: 18500, requests: 1050 },
          widget: { units: 4790, requests: 241 },
        },
        daily: generateDailyData(),
      },
    }),
  },
  {
    method: 'GET',
    pattern: /^\/admin\/usage\/summary$/,
    handler: () => ({
      summary: {
        llm: {
          daily: { used: 342, limit: 1000, percent: 34.2 },
          monthly: { used: 8920, limit: 20000, percent: 44.6 },
        },
        tts: {
          daily: { used: 156, limit: 500, percent: 31.2 },
          monthly: { used: 4230, limit: null, percent: 0 },
        },
        stt: {
          daily: { used: 45, limit: null, percent: 0 },
          monthly: { used: 1120, limit: null, percent: 0 },
        },
      },
    }),
  },
  {
    method: 'GET',
    pattern: /^\/admin\/usage\/limits$/,
    handler: () => {
      initUsageData()
      return { limits: getStore().usageLimits }
    },
  },
  {
    method: 'GET',
    pattern: /^\/admin\/usage\/limits\/([^/]+)\/([^/]+)$/,
    handler: ({ matches }) => {
      initUsageData()
      const limit = getStore().usageLimits.find(
        l => l.service_type === matches[1] && l.limit_type === matches[2]
      )
      return { limit: limit || defaultLimits[0] }
    },
  },
  {
    method: 'POST',
    pattern: /^\/admin\/usage\/limits$/,
    handler: ({ body }) => {
      initUsageData()
      const data = body as Record<string, unknown>
      const limit: UsageLimitData = {
        id: Date.now(),
        service_type: data.service_type as string,
        limit_type: data.limit_type as string,
        limit_value: data.limit_value as number,
        warning_threshold: (data.warning_threshold as number) ?? 80,
        hard_limit: (data.hard_limit as boolean) ?? false,
        enabled: true,
        created: nowISO(),
      }
      getStore().usageLimits.push(limit)
      return { limit }
    },
  },
  {
    method: 'DELETE',
    pattern: /^\/admin\/usage\/limits\/([^/]+)\/([^/]+)$/,
    handler: ({ matches }) => {
      const store = getStore()
      store.usageLimits = store.usageLimits.filter(
        l => !(l.service_type === matches[1] && l.limit_type === matches[2])
      )
      return { status: 'ok' }
    },
  },
  {
    method: 'GET',
    pattern: /^\/admin\/usage\/check\//,
    handler: () => ({
      service_type: 'llm',
      limit_type: 'daily',
      current_usage: 342,
      within_limits: true,
      warnings: [],
      blocked: false,
      limits: [{ limit_type: 'daily', limit_value: 1000, current_usage: 342, usage_percent: 34.2, hard_limit: false }],
    }),
  },
  {
    method: 'POST',
    pattern: /^\/admin\/usage\/cleanup/,
    handler: () => ({ status: 'ok', deleted: 0, retention_days: 90 }),
  },
  {
    method: 'POST',
    pattern: /^\/admin\/usage\/log$/,
    handler: ({ body }) => ({
      log: { id: Date.now(), timestamp: nowISO(), ...(body as object) },
    }),
  },
]

File: app/routers/usage.py
# app/routers/usage.py
"""Usage tracking router - view, configure, and analyze usage logs."""

from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from auth_manager import User, get_current_user, require_admin
from db.database import AsyncSessionLocal
from db.repositories.usage import UsageLimitsRepository, UsageRepository


router = APIRouter(prefix="/admin/usage", tags=["usage"])


# =============================================================================
# Request/Response Models
# =============================================================================


class UsageLogRequest(BaseModel):
    """Request to log a usage event."""

    service_type: str = Field(..., description="Service type: tts, stt, llm")
    action: str = Field(..., description="Action performed: synthesis, transcribe, chat")
    units_consumed: int = Field(1, ge=0, description="Units consumed (chars, tokens, seconds)")
    source: Optional[str] = Field(None, description="Source: admin, telegram, widget")
    source_id: Optional[str] = Field(None, description="Source instance ID")
    cost_usd: Optional[float] = Field(None, ge=0, description="Cost in USD")
    details: Optional[dict] = Field(None, description="Additional details (JSON)")


class UsageLimitRequest(BaseModel):
    """Request to set a usage limit."""

    service_type: str = Field(..., description="Service type: tts, stt, llm")
    limit_type: str = Field(..., description="Limit type: daily, monthly")
    limit_value: int = Field(..., gt=0, description="Limit value in units")
    hard_limit: bool = Field(False, description="Enforce hard limit (block on exceed)")
    warning_threshold: int = Field(80, ge=0, le=100, description="Warning threshold percentage")


# =============================================================================
# Usage Logs Endpoints
# =============================================================================


@router.get("/logs")
async def admin_get_usage_logs(
    service_type: Optional[str] = None,
    action: Optional[str] = None,
    source: Optional[str] = None,
    from_date: Optional[str] = None,
    to_date: Optional[str] = None,
    limit: int = 100,
    offset: int = 0,
    user: User = Depends(get_current_user),
):
    """Get usage logs with optional filters."""
    # Parse dates if provided
    from_dt = None
    to_dt = None
    if from_date:
        try:
            from_dt = datetime.fromisoformat(from_date.replace("Z", "+00:00"))
        except ValueError:
            pass
    if to_date:
        try:
            to_dt = datetime.fromisoformat(to_date.replace("Z", "+00:00"))
        except ValueError:
            pass

    async with AsyncSessionLocal() as session:
        repo = UsageRepository(session)
        logs = await repo.get_usage(
            service_type=service_type,
            action=action,
            source=source,
            from_date=from_dt,
            to_date=to_dt,
            limit=min(limit, 500),  # Cap at 500
            offset=offset,
        )
        return {"logs": logs, "count": len(logs)}


@router.get("/stats")
async def admin_get_usage_stats(
    service_type: Optional[str] = None,
    period_days: int = 30,
    user: User = Depends(get_current_user),
):
    """Get usage statistics for the specified period."""
    if period_days < 1 or period_days > 365:
        raise HTTPException(status_code=400, detail="Period must be between 1 and 365 days")

    async with AsyncSessionLocal() as session:
        repo = UsageRepository(session)
        stats = await repo.get_usage_stats(
            service_type=service_type,
            period_days=period_days,
        )
        return {"stats": stats}


@router.post("/log")
async def admin_log_usage(request: UsageLogRequest):
    """
    Log a usage event.

    This endpoint is called by services (TTS, STT, LLM) to track usage.
    No authentication required for internal service calls.
    """
    if request.service_type not in ("tts", "stt", "llm"):
        raise HTTPException(status_code=400, detail="Invalid service_type. Must be: tts, stt, llm")

    async with AsyncSessionLocal() as session:
        repo = UsageRepository(session)
        log = await repo.log_usage(
            service_type=request.service_type,
            action=request.action,
            units_consumed=request.units_consumed,
            source=request.source,
            source_id=request.source_id,
            cost_usd=request.cost_usd,
            details=request.details,
        )
        return {"log": log}


@router.post("/cleanup")
async def admin_cleanup_usage_logs(
    days: int = 90,
    user: User = Depends(require_admin),
):
    """Delete usage logs older than specified days."""
    if days < 7:
        raise HTTPException(status_code=400, detail="Minimum retention period is 7 days")
    if days > 365:
        raise HTTPException(status_code=400, detail="Maximum retention period is 365 days")

    async with AsyncSessionLocal() as session:
        repo = UsageRepository(session)
        deleted = await repo.cleanup_old_logs(days=days)
        return {"status": "ok", "deleted": deleted, "retention_days": days}


# =============================================================================
# Usage Limits Endpoints
# =============================================================================


@router.get("/limits")
async def admin_get_limits(
    enabled_only: bool = True,
    user: User = Depends(require_admin),
):
    """Get all usage limits."""
    async with AsyncSessionLocal() as session:
        repo = UsageLimitsRepository(session)
        limits = await repo.get_all_limits(enabled_only=enabled_only)
        return {"limits": limits}


@router.get("/limits/{service_type}/{limit_type}")
async def admin_get_limit(
    service_type: str,
    limit_type: str,
    user: User = Depends(require_admin),
):
    """Get a specific usage limit."""
    async with AsyncSessionLocal() as session:
        repo = UsageLimitsRepository(session)
        limit = await repo.get_limit(service_type, limit_type)
        if not limit:
            raise HTTPException(status_code=404, detail="Limit not found")
        return {"limit": limit}


@router.post("/limits")
async def admin_set_limit(
    request: UsageLimitRequest,
    user: User = Depends(require_admin),
):
    """Create or update a usage limit."""
    if request.service_type not in ("tts", "stt", "llm"):
        raise HTTPException(status_code=400, detail="Invalid service_type. Must be: tts, stt, llm")
    if request.limit_type not in ("daily", "monthly"):
        raise HTTPException(status_code=400, detail="Invalid limit_type. Must be: daily, monthly")

    async with AsyncSessionLocal() as session:
        repo = UsageLimitsRepository(session)
        limit = await repo.set_limit(
            service_type=request.service_type,
            limit_type=request.limit_type,
            limit_value=request.limit_value,
            hard_limit=request.hard_limit,
            warning_threshold=request.warning_threshold,
        )
        return {"limit": limit}


@router.delete("/limits/{service_type}/{limit_type}")
async def admin_delete_limit(
    service_type: str,
    limit_type: str,
    user: User = Depends(require_admin),
):
    """Delete a usage limit."""
    async with AsyncSessionLocal() as session:
        repo = UsageLimitsRepository(session)
        deleted = await repo.delete_limit(service_type, limit_type)
        if not deleted:
            raise HTTPException(status_code=404, detail="Limit not found")
        return {"status": "ok"}


# =============================================================================
# Usage Check Endpoint (for services)
# =============================================================================


@router.get("/check/{service_type}")
async def admin_check_usage(
    service_type: str,
    limit_type: str = "daily",
):
    """
    Check if usage is within limits.

    Called by services before processing requests.
    Returns whether the request should proceed or be blocked.
    """
    if service_type not in ("tts", "stt", "llm"):
        raise HTTPException(status_code=400, detail="Invalid service_type")

    async with AsyncSessionLocal() as session:
        usage_repo = UsageRepository(session)
        limits_repo = UsageLimitsRepository(session)

        # Get current usage for period
        current_usage = await usage_repo.get_period_usage(service_type, limit_type)

        # Check against limits
        check_result = await limits_repo.check_limit(service_type, current_usage)

        return {
            "service_type": service_type,
            "limit_type": limit_type,
            "current_usage": current_usage,
            **check_result,
        }


@router.get("/summary")
async def admin_get_usage_summary(
    user: User = Depends(get_current_user),
):
    """
    Get a summary of usage across all services.

    Includes current usage vs limits for quick dashboard display.
    """
    async with AsyncSessionLocal() as session:
        usage_repo = UsageRepository(session)
        limits_repo = UsageLimitsRepository(session)

        summary = {}
        for service in ("tts", "stt", "llm"):
            daily_usage = await usage_repo.get_period_usage(service, "daily")
            monthly_usage = await usage_repo.get_period_usage(service, "monthly")

            daily_limit = await limits_repo.get_limit(service, "daily")
            monthly_limit = await limits_repo.get_limit(service, "monthly")

            summary[service] = {
                "daily": {
                    "used": daily_usage,
                    "limit": daily_limit["limit_value"] if daily_limit else None,
                    "percent": round(daily_usage / daily_limit["limit_value"] * 100, 1)
                    if daily_limit and daily_limit["limit_value"] > 0
                    else 0,
                },
                "monthly": {
                    "used": monthly_usage,
                    "limit": monthly_limit["limit_value"] if monthly_limit else None,
                    "percent": round(monthly_usage / monthly_limit["limit_value"] * 100, 1)
                    if monthly_limit and monthly_limit["limit_value"] > 0
                    else 0,
                },
            }

        return {"summary": summary}
